src/flow/src/batching_mode/engine.rs

```diff
@@ -0,0 +1,306 @@
+use common_telemetry::info;
+use common_telemetry::tracing::warn;
+use query::QueryEngineRef;
+use snafu::{ensure, OptionExt, ResultExt};
+use store_api::storage::RegionId;
+use table::metadata::TableId;
+use tokio::sync::{oneshot, RwLock};
+
+use super::frontend_client::FrontendClient;
+use crate::adapter::{CreateFlowArgs, FlowId, TableName};
+use crate::batching_mode::task::BatchingTask;
+use crate::batching_mode::time_window::{find_time_window_expr, TimeWindowExpr};
+use crate::batching_mode::utils::sql_to_df_plan;
+use crate::error::{ExternalSnafu, FlowAlreadyExistSnafu, TableNotFoundMetaSnafu, UnexpectedSnafu};
+use crate::Error;
+
+/// Batching mode Engine, responsible for driving all the batching mode tasks
+///
+/// TODO(discord9): determine how to configure refresh rate
+pub struct BatchingEngine {
+    tasks: RwLock<BTreeMap<FlowId, BatchingTask>>,
+    shutdown_txs: RwLock<BTreeMap<FlowId, oneshot::Sender<()>>>,
+    frontend_client: Arc<FrontendClient>,
+    flow_metadata_manager: FlowMetadataManagerRef,
+    table_meta: TableMetadataManagerRef,
+    query_engine: QueryEngineRef,
+}
+
+impl BatchingEngine {
+    pub fn new(
+        frontend_client: Arc<FrontendClient>,
+        query_engine: QueryEngineRef,
+        flow_metadata_manager: FlowMetadataManagerRef,
+        table_meta: TableMetadataManagerRef,
+    ) -> Self {
+        Self {
+            tasks: Default::default(),
+            shutdown_txs: Default::default(),
+            frontend_client,
+            flow_metadata_manager,
+            table_meta,
+            query_engine,
+        }
+    }
+
+    pub async fn handle_inserts(
+        &self,
+        request: api::v1::region::InsertRequests,
+    ) -> Result<FlowResponse, Error> {
+        let table_info_mgr = self.table_meta.table_info_manager();
+        let mut group_by_table_id: HashMap<TableId, Vec<api::v1::Rows>> = HashMap::new();
+
+        for r in request.requests {
+            let tid = RegionId::from(r.region_id).table_id();
+            let entry = group_by_table_id.entry(tid).or_default();
+            if let Some(rows) = r.rows {
+                entry.push(rows);
+            }
+        }
+
+        let tids = group_by_table_id.keys().cloned().collect::<Vec<TableId>>();
+        let table_infos =
+            table_info_mgr
+                .batch_get(&tids)
+                .await
+                .with_context(|_| TableNotFoundMetaSnafu {
+                    msg: format!("Failed to get table info for table ids: {:?}", tids),
+                })?;
+
+        if !tids.iter().all(|id| table_infos.contains_key(id)) {
+            warn!(
+                "Failed to get all the table info for table ids, expected table ids: {:?}, actual table infos' table id: {:?}, those table doesn't exist: {:?}",
+                tids,
+                table_infos.keys(),
+                tids.iter().filter(|id| !table_infos.contains_key(id)).collect::<Vec<_>>()
+            );
```

Comment:
nit: make it more concise
```suggestion
            warn!(
                "Failed to get all the table info for table ids, expected table ids: {:?}, those table doesn't exist: {:?}",
                tids,
                tids.iter().filter(|id| !table_infos.contains_key(id)).collect::<Vec<_>>()
            );
```

Reply:
And the missing table ids is calculated twice in this case